### PR TITLE
[SPARK-34187][SS] Use available offset range obtained during polling when checking offset validation

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch uses the available offset range obtained during polling Kafka records to do offset validation check.

### Why are the changes needed?

We support non-consecutive offsets for Kafka since 2.4.0. In `fetchRecord`, we do offset validation by checking if the offset is in available offset range. But currently we obtain latest available offset range to do the check. It looks not correct as the available offset range could be changed during the batch, so the available offset range is different than the one when we polling the records from Kafka.

It is possible that an offset is valid when polling, but at the time we do the above check, it is out of latest available offset range. We will wrongly consider it as data loss case and fail the query or drop the record.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

This should pass existing unit tests.

This is hard to have unit test as the Kafka producer and the consumer is asynchronous. Further, we also need to make the offset out of new available offset range.

### CODE REVIEW

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/consumer/KafkaDataConsumer.scala [hunk 1]
```diff
@@ -192,6 +197,13 @@ private[consumer] case class FetchedData(
    * Returns the next offset to poll after draining the pre-fetched records.
    */
   def offsetAfterPoll: Long = _offsetAfterPoll
+
+  /**
+   * Returns the tuple of earliest and latest offset that are the available offset range when
```
Comment (jaceklaskowski) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/consumer/KafkaDataConsumer.scala:202]: nit: offsets (plural) and s/that are/that is (singular)

**Comment by HyukjinKwon** [comment]
cc @HeartSaVioR @gaborgsomogyi @xuanyuanking

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/consumer/KafkaDataConsumer.scala [hunk 2]
```diff
@@ -192,6 +197,13 @@ private[consumer] case class FetchedData(
    * Returns the next offset to poll after draining the pre-fetched records.
    */
   def offsetAfterPoll: Long = _offsetAfterPoll
+
+  /**
+   * Returns the tuple of earliest and latest offsets that is the available offset range when
```
Comment (jaceklaskowski) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/consumer/KafkaDataConsumer.scala:202]: nit: Use `@returns` annotation.

Sorry for being late with this, but I've just noticed it.

**Review by dongjoon-hyun — COMMENTED** [review]
Given your assessment, could you make it as a correctness issue, @viirya ?
> It looks not correct ...
> ... drop the record.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/consumer/FetchedDataPool.scala [hunk 3]
```diff
@@ -170,11 +170,14 @@ private[consumer] object FetchedDataPool {
   }
 
   private object CachedFetchedData {
+    import KafkaDataConsumer.AvailableOffsetRange
```
Comment (dongjoon-hyun) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/consumer/FetchedDataPool.scala:173]: Shall we move this to line 31?

**Review by dongjoon-hyun — APPROVED** [review]
Although we don't have a test case, this looks reasonable. I left a few minor comments.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/consumer/KafkaDataConsumer.scala (hunk 2 above)
Comment (viirya) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/consumer/KafkaDataConsumer.scala:202]: Hm, I think this just follows other methods above and below.

**Comment by viirya** [comment]
> Given your assessment, could you make it as a correctness issue, @viirya ?

Done. Thanks.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/consumer/FetchedDataPool.scala (hunk 3 above)
Comment (viirya) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/consumer/FetchedDataPool.scala]: Moved. Thanks.

**Comment by dongjoon-hyun** [comment]
Thank you for update, @viirya .

**Comment by dongjoon-hyun** [comment]
Merged to master/3.1. Could you make a backport to 3.0/2.4?

**Comment by viirya** [comment]
Thanks @dongjoon-hyun. Will file backport later.

**Review by gaborgsomogyi — COMMENTED** [review]
Late LGTM. It took me some time to reproduce it on cluster but it works like charm.

**Comment by HeartSaVioR** [comment]
Nice finding, and thanks for the fix!

**Pull request closed without merging**